[LLVM][Intrinsics] Adds an API to automatically resolve overload types

rajatbajpai opened the pull request:
Currently, the getOrInsertDeclaration API requires callers to explicitly provide overload types for overloaded intrinsics, placing a significant burden on callers who must determine whether overload types are needed. This typically results in conditional logic at each call site to check if the intrinsic is overloaded and manually match the intrinsic signature.

This patch introduces a new getOrInsertDeclaration overload that automatically deduces overload types from the provided return type and argument types, then uses this API to simplify IRBuilder::CreateIntrinsic. The new API uses Intrinsic::matchIntrinsicSignature internally to resolve overloaded types, eliminating the need for callers to do manual overload detection.

rajatbajpai committed:
[LLVM][Intrinsics] Adds an API to automatically resolve overload types

Currently, the getOrInsertDeclaration API requires callers to explicitly
provide overload types for overloaded intrinsics, placing a significant burden
on callers who must determine whether overload types are needed. This typically
results in conditional logic at each call site to check if the intrinsic is
overloaded and manually match the intrinsic signature.

This patch introduces a new getOrInsertDeclaration overload that automatically
deduces overload types from the provided return type and argument types, then
uses this API to simplify IRBuilder::CreateIntrinsic. The new API uses
Intrinsic::matchIntrinsicSignature internally to resolve overloaded types,
eliminating the need for callers to do manual overload detection.

durga4github commented on llvm/unittests/IR/IntrinsicsTest.cpp:
```diff
@@ -197,4 +197,169 @@ TEST(IntrinsicAttributes, TestGetFnAttributesBug) {
+  ArgTys32.push_back(Type::getInt32Ty(Context));
+  ArgTys32.push_back(Type::getInt32Ty(Context));
+
+  Function *F32 = Intrinsic::getOrInsertDeclaration(M.get(), Intrinsic::umax,
```
ok, I almost read it as f32/f64 for float/double..

Can we name it as Func32 or FuncI32 or something like that, so that it is clear that we are looking at Integer types here?

jurahul commented on llvm/lib/IR/Intrinsics.cpp:
```diff
@@ -746,11 +746,49 @@ Function *Intrinsic::getOrInsertDeclaration(Module *M, ID id,
+  return getOrInsertIntrinsicDeclarationImpl(M, id, Tys, FT);
+}
+
+Function *Intrinsic::getOrInsertDeclaration(Module *M, ID IID, Type *RetTy,
```
nit: use name `id` consistent with the rest of these functions. We can rename all in a later PR.

rajatbajpai committed:
Addressed review comments

1. Replaced `IID` with `id` for intrinsic ID.
2. Fixed variable names.
3. Added IR test cases which are overloaded on adress space.

durga4github reviewed (approved):
LGTM

rajatbajpai committed:
Addressed review comment

Updated comment to include value as well.

rajatbajpai commented:
Changes have been merged. Thanks everyone for taking a time to review this change.
